### PR TITLE
file: ensure the parent directory exists if the open mode contains the 'create' or 'write' flag

### LINKED ISSUE #1984: Deleted shaderlog dir doesn't get recreated in OGL
DX12 recreates the directory if you delete it, while OGL just throws this error:

> F {rsx::thread} class std::logic_error thrown: fs::file is null

PS: idk how Vulkan behaves, i can't test it.

### PR DESCRIPTION
**Pull request by GhettoGirl**
Fixes #1984 and #1985

### CODE REVIEW

**Comment by Nezarn** [comment]
Yep, it fixes it

**Comment by GhettoGirl** [comment]
also fixes ```fs:file is null``` on Linux because of missing **shaderlog** directory in ```$XDG_CONFIG_DIR/rpcs3```.

**Comment by raven02** [comment]
Humm just tried this build with Bomberman Ultra but still see the fs::file null error. However not too sure if it is related to shaderlog.

F {PPU[0x64] Thread (ReadProfileSettings) [0x00162960]} class std::logic_error thrown: fs::file is null

**Comment by GhettoGirl** [comment]
@raven02 Than is may be something else than. I just call ```fs::create_path()``` to ensure the shaderlog folder exists, because it doesn't gets created when deleted.
In Linux and Mac this folder usually doesn't exists in the config directory (```fs::get_config_dir()```), only on Windows because its part of the repository. I'm not sure if this is the best location for the call tho :confused:

**Comment by raven02** [comment]
Yep, i'll get some time to investigate it.

**Comment by paulsapps** [comment]
@GhettoGirl "procmon" is like "strace" for windows

**Comment by kd-11** [comment]
By the way, vulkan also does not check to recreate the directory. You can find the log writes in vkvertexprogram.cpp and vkfragmentprogram.cpp.

**Comment by GhettoGirl** [comment]
Added *fs::create_path* calls in Vulkan.

File: rpcs3/Emu/RSX/VK/VKFragmentProgram.cpp [hunk 1]
```diff
@@ -371,6 +371,7 @@ void VKFragmentProgram::Decompile(const RSXFragmentProgram& prog)
 
 void VKFragmentProgram::Compile()
 {
+	fs::create_path(fs::get_config_dir() + "/shaderlog");
 	fs::file(fs::get_config_dir() + "shaderlog/FragmentProgram.spirv", fs::rewrite).write(shader);
```
Comment (paulsapps) [inline rpcs3/Emu/RSX/VK/VKFragmentProgram.cpp]: Is there not an fs::xyz to say create the path to the file?

Comment (GhettoGirl) [inline rpcs3/Emu/RSX/VK/VKFragmentProgram.cpp]: I don't thinks so, but extending the fs::file class when the fs::rewrite is passed should create the path too, not just the file.
Than there is no need to call the create_path function everytime.
But thats my opinion.

**Comment by GhettoGirl** [comment]
I had an idea and added an enhancement into ```file::open(string, open_mode)```. There is now a check if the open_mode contains the flags **create** or **write** and if the parent directory exists where the file should be written. If the so called directory doesn't exists or if its an file for some reason, an error is logged so that the user knows whats going on.

This should avoid unnecessary *fs::create_path* everywhere around the emulator.

Maybe needs performance testing.

**Comment by Nekotekina** [comment]
Aww, this is wrong. Creating directory is not expected from fs::open, at least not in this way.

**Comment by GhettoGirl** [comment]
I've reverted the last commit.

**Comment by Nekotekina** [comment]
Just create /shaderlog/ in Emulator::Init or somewhere else...
Although I wanted to switch to "get_data_dir" for this purpose.

**Comment by GhettoGirl** [comment]
I will look into this another time and use the get_data_dir function.

**Pull request merged**